cantusdb/turtle/csv2rdf_single_subject.py

```diff
@@ -0,0 +1,67 @@
+from rdflib import Graph, URIRef, Literal
+from rdflib.namespace import RDF
+import csv
+import validators
+import sys
+import json
+
+
+# a main operation function
+def convert_csv_to_turtle(filename) -> Graph:
+    """
+    Adds all informations as RDF triples from the input filename into a graph and return it.
+    @Pre: input filename must be type string.
+    @Post: Returns a RDF.Graph that has all the triples
+    """
+
+    if len(sys.argv) != 2:
+        raise ValueError("Invalid number of input filename")
+
+    with open(filename, "r", encoding="utf-8") as csv_file:
+        # TODO: maybe change this one to accomadate for multiple files
+
+        g = Graph()
+
+        csv_reader = csv.reader(csv_file)
+        with open("relations_mapping.json", "r") as mapper:
+            ontology_dict = json.load(mapper)
+
+        header = next(csv_reader)
+
+        # Convert each row to Turtle format and add it to the output
+        for row in csv_reader:
+            # the first column as the subject
+            key_attribute = URIRef(row[0])
+            if "type" in ontology_dict:
+                g.add((key_attribute, RDF.type, URIRef(ontology_dict["type"])))
+
+            # extracting other informations
+            # TODO: is the source description texts needed?
+            for i, element in enumerate(row[1:]):
+                predicate_cur = header[i + 1]
```

Comment:
As above, anything you are doing with the header I think could be done once, before iterating through all of the csv rows.

Reply:
For every row in the csv, you are:
- slicing the `header` list to obtain the value of the header for that row (calling this `predicate_cur`)
- looking for whether or not `predicate_cur` is in `ontology_dict` (line 49)
- creating a URIRef based on that `predicate_cur` (line 52)

All of those steps are they exact same for each row, right? 

For every row, `header[i]` will be the same, so for every row, `header[i]` will always be in `ontology_dict` or it won't, and so the URIRef you create for a specific column will be the same for every row. 

So then I wonder why you are having your code do the same thing over and over again.